### PR TITLE
Added new plugin Potential Energy Surface Sampler

### PR DESCRIPTION
**Pull request by ml-evs**
This PR adds new plugin functionality to the Enthought examples repo. The Potential Energy Surface Sampler plugin introduces the following classes (and associated factories/models):

- `RandomSamplingMCO`: randomly sample from a DataSource with an arbitrary number of dimensions.
- `EggboxPESDataSource`: produce a random n-dimensional Eggbox potential energy surface [1] via a square lattice of "basins" and their depths. Increasing the variance of basin depths, σ*, allows deeper basins to swallow up the shallower ones, mimicking an atomistic PES (see attached plots, created with
`scripts/plot_sigma.py`). The `RandomSamplingMCO` provides trial points to this data source at which to (optionally) locally optimise the potential; the data source then returns the position and value of the potential at the mininum.

Still to-do:
- ~implement PBCs for 2D and beyond case, by defining the lattice vectors. This should allow for arbitrary lattices to be constructed (see e.g. bleeding in the corners of plots)~ done.
- ~maintain a list of "discovered" basins in the MCO model.~ not going to do.

pot_2d

[1] "Exploring the origins of the power-law properties of energy landscapes: An egg-box model", Claire Massen *et al*. (2005) https://arxiv.org/abs/cond-mat/0612205

### CODE REVIEW

**Comment by ml-evs** [comment]
This PR should be complete, minus the caveats listed in the to-do above, which can be added in a subsequent PR.

File: eggbox_potential_sampler/eggbox_pes_data_source/tests/test_data_source.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def test_1d_run(self):
+        ds = EggboxPESDataSource(self.factory)
+        model = EggboxPESDataSourceModel(self.factory)
+        model.num_cells = 5
+        model.dimension = 1
+        model.sigma_star = 0.1
+
+        i = 0
+        while i < 10:
+            random = np.random.rand()
```
Comment (jjenthought) [inline eggbox_potential_sampler/eggbox_pes_data_source/tests/test_data_source.py]: I think as a general rule it's good to avoid randomness in tests so you don't get intermittent test failures. I get that you may be able to identify more corner cases by sampling a wider space, but ideally these corner cases would be something you test for directly

Comment (jjenthought) [reply eggbox_potential_sampler/eggbox_pes_data_source/tests/test_data_source.py]: I guess on the back of that comment, running the test multiple times may not be needed if you choose to go with a non-random value

File: eggbox_potential_sampler/random_sampling_mco/tests/test_example_mco_factory.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import unittest
+
+# from force_bdss.core_plugins import BasicMCOCommunicator
```
Comment (jjenthought) [inline eggbox_potential_sampler/random_sampling_mco/tests/test_example_mco_factory.py]: Commented code

File: eggbox_potential_sampler/random_sampling_mco/tests/test_example_mco_factory.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+from enthought_example.example_mco.example_mco_communicator import \
```
Comment (jjenthought) [inline eggbox_potential_sampler/random_sampling_mco/tests/test_example_mco_factory.py]: We should probably go with the bracketed list of imports since we have one a few lines down. I think they recommend doing that even for lists of 1!

**Comment by ml-evs** [comment]
Potentially something to be investigated in how the BDSS handles `PLUGIN_VERSION` (currently expects PositiveInt), but think its safe to merge this for now.
